Put the snippets used in doc in their own folder

skyluc opened the pull request:
### What changes were proposed in this pull request?
Extract the code snippets used in the documentation to the `example/snippets` folder.
Some of the code snippets were already taken from the projects in `examples`, other were hardcoded in the documentation.
The snippets are now in projects in the `snippets` folder. They are compiled and tested with the other example projects.

This only part of the code snippets to extract from the doc, but the structure change in the documentation is needed for other reasons.

### Why are the changes needed?
The code snippets in the documentation need to be compiled and check, to make sure that they are always compatible with the core project.


### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Running the documentation build, with equivalent PR, checking the result.

skyluc commented:
related cloudflow-docs PR lightbend/cloudflow-docs#40

agemooij commented on examples/examples.yaml:
```diff
@@ -21,3 +21,35 @@ examples:
   - id: "tensorflow-akka"
     path: "tensorflow-akka"
     description: This app shows how to serve a Tensorflow model using Akka
+  - id: "sensor-data-scala-snippets"
```
@skyluc we might have to re-assess the usage of this file. It had a long-forgotten purpose at some point but it would be interesting to see whether we still need a manually updated list of example projects.

skyluc replied:
It is used in `scripts/build-all.sh`, which is used in the PR check to build and test all example projects.
It is really practical, I only had to add the new projects to the list, and they became part of the CI.

agemooij replied:
Wouldn't it work just as well to just list the number of sub-dirs or `find` all sbt projects in the structure?

agemooij commented on examples/examples.yaml:
```diff
@@ -21,3 +21,35 @@ examples:
+  - id: "build-akka-streamlets-java-snippets"
+    path: "snippets/modules/ROOT/examples/build-akka-streamlets-java"
+    description: |
+      Set of files containing the Scala code snippets used in the build-akka-streamlets documentation
```
Is there a good reason to keep this separate from `akkastreams-java` (and the same for the scala side)? Whot not have less top-level containers and (if/when necessary) just use folders to structure the contents of such projects.

I had expected one single snippets project or perhaps, if that made more sense, one per runtime. I'm not sure this additional directory structure serves a good purpose.

agemooij replied:
Every extra sbt project we add here requires us to updates its dependencies, etc. I'd prefer the smallest possible number of snippet projects and then use internal directories to add structure where needed. If the Java vs Scala difference is tricky then that could be a necessary reason for having more than 1 snippets (e.g. 2) projects.

skyluc replied:
One of the reasons to have different project is that most of them define a blueprint
And I remember that we had problems with sbt projects defining multiple Cloudflow app subproject, and we include blueprint in the documentation, which have to be verified.
So I went the easy way, one project per blueprint related bloc of snippets.

Also, having one, or a couple of, big sbt project will only be moving the complexity.
The `build-*-streamlets-*` projects already contain each 4 library projects, one for each steps of the explanation, and one app project for the blueprint.

But I should be able to group some of the simple projects.

agemooij replied:
ok, thanks. Just trying to keep a handle on the complexity 😉

agemooij commented on examples/snippets/antora.yml:
```diff
@@ -0,0 +1,3 @@
+name: "docsnippets"
+version: 1.3.1
```
Do we need the version hardcoded here?
Every additional place that hardcodes our version number is one more problem to solve for automated release management.

skyluc replied:
Yes, we will need a specific version number, to support multi-version documentation
On master, it can be `latest` on something like that. But in the branches, it should be the actual version.

agemooij replied:
How can we keep this maintainable? 
Should this be auto-updated by the release automation?

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/akkastreams-java/project/cloudflow-plugins.sbt:
```diff
@@ -0,0 +1,5 @@
+//
+resolvers += Resolver.url("cloudflow", url("https://lightbend.bintray.com/cloudflow"))(Resolver.ivyStylePatterns)
+
+addSbtPlugin("com.lightbend.cloudflow" % "sbt-cloudflow" % "1.3.1-RC4")
```
This should be 1.3.1?

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/akkastreams-java/src/main/java/cloudflow/akkastreamsdoc/FilterStreamlet.java:
```diff
@@ -0,0 +1,129 @@
+import cloudflow.streamlets.avro.AvroInlet;
+import cloudflow.streamlets.avro.AvroOutlet;
+
+import java.nio.file.Path;
```
[minor] we usually put the java imports first, then scala, then ...

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/akkastreams-scala/build.sbt:
```diff
@@ -0,0 +1,42 @@
+        "com.lightbend.akka"     %% "akka-stream-alpakka-file"  % "1.1.2",
+        "com.typesafe.akka"      %% "akka-http-spray-json"      % "10.1.10",
+        "ch.qos.logback"         %  "logback-classic"           % "1.2.3",
+        "com.typesafe.akka"      %% "akka-http-testkit"         % "10.1.10" % "test",
```
Have you checked, for instance by running these examples, that it does not start with many versions of Akka libraries on classpath? We depend on `val AkkaHttp     = "10.1.9"` in cloudflow itself. Probably best to keep that the same.

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/akkastreams-scala/build.sbt:
```diff
@@ -0,0 +1,42 @@
+    .settings(
+//end::docs-projectSetup-example[]
+      libraryDependencies ++= Seq(
+        "com.lightbend.akka"     %% "akka-stream-alpakka-file"  % "1.1.2",
```
This might depend on a different Akka version than `2.5.29`

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/akkastreams-scala/project/cloudflow-plugins.sbt:
```diff
@@ -0,0 +1,5 @@
+//
+resolvers += Resolver.url("cloudflow", url("https://lightbend.bintray.com/cloudflow"))(Resolver.ivyStylePatterns)
+
+addSbtPlugin("com.lightbend.cloudflow" % "sbt-cloudflow" % "1.3.1-RC4")
```
```suggestion
addSbtPlugin("com.lightbend.cloudflow" % "sbt-cloudflow" % "1.3.1")
```

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/akkastreams-scala/src/main/scala/cloudflow/akkastreamsdoc/DataHttpIngress.scala:
```diff
@@ -0,0 +1,18 @@
+import JsonSupport._
+
+class DataHttpIngress extends AkkaServerStreamlet {
+  val out         = AvroOutlet[Data]("out").withPartitioner(RoundRobinPartitioner)
```
very minor, you can leave the `.withPartitioner(RoundRobinPartitioner)` out cause it is default. (maybe for didactic purposes it's better to be explicit so not sure)

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/build-akka-streamlets-java/step0/src/main/java/com/example/ReportPrinter.java:
```diff
@@ -0,0 +1,18 @@
+  public StreamletShape shape() { throw new UnsupportedOperationException("Not Implemented"); }
+  // 3. TODO Override createLogic to provide StreamletLogic
+  public RunnableGraphStreamletLogic createLogic() { throw new UnsupportedOperationException("Not Implemented"); }
+}
```
!WHY LUC WHY!
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206
lol I'm just kidding of course.

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/sensor-data-scala/src/main/scala/sensordata/ConfigParams.scala:
```diff
@@ -0,0 +1,36 @@
+  )
+
+  /*
+  val InfluxDBActive = BooleanConfigParameter(
```
Remove commented code please.

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/sensor-data-scala/src/main/scala/sensordata/ConfigParams.scala:
```diff
@@ -0,0 +1,36 @@
+
+import cloudflow.streamlets._
+
+trait ConfigParams {
```
Its not clear whose configparams these are.. I would prefer to define these where they are used.

RayRoestenburg replied:
Oh it also looks like (unless there is a 'too big diff' in here) that it is not used?

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/sensor-data-scala/.gitignore:
```diff
@@ -0,0 +1,30 @@
+etc/bootstrap-local.conf
```
Why is the entire sensor-data-scala copied in here @skyluc ? Can we avoid it?

agemooij replied:
As discussed offline, the idea is that eventually we get rid of the standalone sensor-data-xxx examples in favor of these snippets. The sensor data examples were always meant as getting started info more than full examples.

RayRoestenburg commented on examples/snippets/modules/ROOT/examples/build-akka-streamlets-scala/build.sbt:
```diff
@@ -0,0 +1,75 @@
+lazy val root =
```
The build akka streamlets example looks really nice, good to have it added!

agemooij commented:
> I'm not sure what exactly was discussed but would be great not to duplicate sensor-data-scala code. What is the plan?

As discussed offline, the idea is that eventually we get rid of the standalone sensor-data-xxx examples in favor of these snippets. The sensor data examples were always meant as getting started info more than full examples.

skyluc commented:
@RayRoestenburg the goal of this PR is to start transferring the base files used for code snippets in the documentation, to a location where Antora can easily access then, and they can also be shared with the enterprise documentation.
I didn't not create any new documentation, and did not fix any of the code snippets, except to make sure that they actually compile.

As for the copy of `sensor-data-scala`, I would have loved to do without it, but it is use as the base of the `develop-example-streamlets` documentation page.

@agemooij if `sensor-data-scala` is not meant to stay in the 'normal' examples, I think we should move it in the snippets section now. It would be a cleaner transition.

agemooij commented:
> @agemooij if `sensor-data-scala` is not meant to stay in the 'normal' examples, I think we should move it in the snippets section now. It would be a cleaner transition.

Agreed. Feel free to do that as part of this PR. Both existing `sensor-data-...` example projects can be removed AFAIC

skyluc commented:
Moved `sensor-data-scala`, and remove `sensor-data-java`.
Updated the version numbers for the cloudflow plugin.
Set a generic version for the snippets (until we correctly document how version will work).

I will likely merge this PR tomorrow morning, so we can also merge the `cloudflow-docs` one.

skyluc commented:
Merging this PR, to be able to merge the `cloudflow-docs` one.
I'll go through the last comments in the next PR.
